src/components/Editor/CitationManager/index.js

```diff
@@ -0,0 +1,333 @@
+    }
+
+    citationInfo.addNode(node)
+    this.sortList()
+
+    return new Proxy(citationInfo, CitationInfoHandler)
+  }
+
+  /**
+   * disconnect node from citation
+   * @param {Proxy} citation proxy of citationInfo
+   * @param {Node} node citation node in tiptap editor
+   */
+  unregisterNode(citation, node) {
+    const { title, url } = citation.info
+    const key = `${title}-${url}`
+
+    const citationInfo = this.citationMap.get(key)
+
+    if (!citationInfo) {
+      return
+    }
+
+    const citationNode = new CitationNode(node)
+    const nodeListEmpty = citationInfo.removeNode(citationNode)
+    if (nodeListEmpty) {
+      this.citationList.splice(citationInfo.index, 1)
+      this.citationMap.delete(key)
+      this.updateCitationIndex()
+    }
+  }
+
+  /**
+   * update citation info
+   * @param {Proxy} citation proxy of citationInfo
+   * @param {String} title new title
+   * @param {String} url new url
+   */
+  updateCitation(citation, title, url) {
+    const { title: oldTitle, url: oldUrl } = citation.info
+    const key = `${oldTitle}-${oldUrl}`
+
+    const citationInfo = this.citationMap.get(key)
+
+    if (!citationInfo) {
+      throw new CitationOpError('CitationManager: Citation not existed')
+    }
+
+    citationInfo.updateInfo({ title, url })
+
+    const newKey = `${title}-${url}`
+
+    this.citationMap.delete(key)
+    this.citationMap.set(newKey, citationInfo)
+  }
+
+  /**
+   * remove citation including all nodes in the post
+   * @param {Proxy} citation proxy of citationInfo
+   */
+  removeCitation(citation) {
+    const { title, url } = citation.info
+    const key = `${title}-${url}`
+
+    const citationInfo = this.citationMap.get(key)
+
+    if (!citationInfo) {
+      throw new CitationOpError('CitationManager: Citation not existed')
+    }
+
+    citationInfo.removeAllNode()
+  }
+
+  /**
+   * @returns {Array<Proxy>} list of proxy of citationInfo
+   */
+  getList() {
+    return this.citationList.map(
+      (citationInfo) => new Proxy(citationInfo, CitationInfoHandler)
+    )
+  }
+
+  getPublishList() {
+    return this.citationList.map((citationInfo) => citationInfo.info)
+  }
+
+  /**
+   * sort citation list from top to bottom by last node of each citation
+   */
+  async sortList() {
+    const blocks = store.state.post.blocks.map((b) => b.id)
+
+    // create an object with key of editor ids
+    const bucket = blocks.reduce((acc, curr) => ({ ...acc, [curr]: [] }), {})
+
+    // put citation into bucket groupped by editorId
+    for (const citationInfo of this.citationList) {
+      bucket[citationInfo.nodes[0].editorId].push(citationInfo)
+    }
+
+    // sort by position in editor block
+    for (const [k, v] of Object.entries(bucket)) {
+      bucket[k] = v.sort((a, b) => a.nodes[0].pos > b.nodes[0].pos)
+    }
+
+    // set sorted list in citationList
+    this.citationList = Object.values(bucket).flatMap((b) => b)
+
+    await this.updateCitationIndex()
```

Comment:
What is the reason using async here? Since the code seems to be synchronous function.

Reply:
removed.